### PR TITLE
[ShaderGraph] [2022.2] [Sapphire] Fix for 1387468 - ShaderGraph screen coordinates work with URP Render Scale

### PR DESCRIPTION
**Pull request by cdxntchou**
### Purpose of this PR
Fix for https://fogbugz.unity3d.com/f/cases/1387468/

Backports:
2022.2: https://github.com/Unity-Technologies/Graphics/pull/6969
2022.1: https://github.com/Unity-Technologies/Graphics/pull/6625

When using a dynamic resolution, the ShaderGraph conversion from pixel coordinate to screen position [0..1] is incorrect, as it uses _ScreenParams, which does not reflect the dynamic resolution.

This PR changes it to use _ScaledScreenParams instead, which will reflect the dynamic resolution.

Tests are also added to ensure this functionality continue to work in the static and dynamic resolution use cases.
It tests raw values returned by the screen and pixel positions, as well as the depth and frame buffer read nodes, and also shows the raw values returned by _ScreenParams and _ScaledScreenParams.
image

---
### Testing status
Added tests to verify the behavior of pixel and screen coordinates in ShaderGraph for URP running with static and dynamic resolution.

This PR:
PR Smoke: 🟢
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252F1387468-master/.yamato%252Fall-smoke_test.yml%2523PR_Smoke_Test_trunk/11490547/job

PR Universal: 🟢
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252F1387468-master/.yamato%252Fall-universal_split.yml%2523PR_Universal_Split_trunk/11470014/job/pipeline

PR URP Foundation: 🟢
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252F1387468-master/.yamato%252Fall-urp_foundation.yml%2523PR_URP_Foundation_trunk/11424635/job/pipeline

Extra URP Foundation: 🟢
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252F1387468-master/.yamato%252Fall-urp_foundation.yml%2523Extra_URP_Foundation_trunk/11464774/job/pipeline

PR ShaderGraph: 🟡 -- failures match master (missing docs - core)
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252F1387468-master/.yamato%252Fall-shadergraph.yml%2523PR_ShaderGraph_trunk/11499559/job/pipeline

Master state:   de54dcca800f33ce8bca5a1398b800047a675e94
   Smoke DX11: 🟢
   PR ShaderGraph: 🔴 only failing missing docs - core
   PR URP: 🟢

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

File: com.unity.shadergraph/Editor/Generation/Targets/BuiltIn/ShaderLibrary/ShaderVariablesFunctions.hlsl [hunk 1]
```diff
@@ -40,7 +40,8 @@ VertexNormalInputs GetVertexNormalInputs(float3 normalOS, float4 tangentOS)
 
 float4 GetScaledScreenParams()
 {
-    return _ScaledScreenParams;
+    // NOTE: Builtin does not use _ScaledScreenParams, so we simply return _ScreenParams instead
+    return _ScreenParams;
```
Comment (jessebarker) [inline com.unity.shadergraph/Editor/Generation/Targets/BuiltIn/ShaderLibrary/ShaderVariablesFunctions.hlsl]: Thanks for the comment here.  This kind of info gets lost easily (or is never found).

File: TestProjects/UniversalGraphicsTest_Foundation/Assets/ControlRenderScale.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+public class ControlRenderScale : MonoBehaviour
+{
+    public float playRenderScale = 0.25f;
+    float originalScale = 1.0f;
+
+    // Update is called once per frame
+    void Update()
+    {
+        // this doesn't seem to do anything...
+        // ScalableBufferManager.ResizeBuffers(0.25f, 0.25f);
```
Comment (jessebarker) [inline TestProjects/UniversalGraphicsTest_Foundation/Assets/ControlRenderScale.cs:17]: Do we need this comment here?  Or rather, what is it telling the next person who works on the code?

File: TestProjects/UniversalGraphicsTest_Foundation/Assets/ControlRenderScale.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    private void OnEnable()
+    {
+        // UniversalRenderPipelineAsset asset = new UniversalRenderPipelineAsset();
```
Comment (jessebarker) [inline TestProjects/UniversalGraphicsTest_Foundation/Assets/ControlRenderScale.cs:22]: Can this be removed?

**Review by jessebarker — APPROVED** [review]
A couple of small nits, but overall it looks fine to me.  Wondering if your SGNumbers stuff couldn't be deployed more widely in shader graph based tests across the Graphics packages.

File: TestProjects/UniversalGraphicsTest_Foundation/Assets/ControlRenderScale.cs (hunk 2 above)
Comment (cdxntchou) [reply TestProjects/UniversalGraphicsTest_Foundation/Assets/ControlRenderScale.cs:17]: Ah, yeah, these should probably just be deleted.

Comment (cdxntchou) [reply TestProjects/UniversalGraphicsTest_Foundation/Assets/ControlRenderScale.cs]: removed the comments

File: TestProjects/UniversalGraphicsTest_Foundation/Assets/ControlRenderScale.cs (hunk 3 above)
Comment (cdxntchou) [reply TestProjects/UniversalGraphicsTest_Foundation/Assets/ControlRenderScale.cs]: yup!  Removed

Comment (cdxntchou) [reply TestProjects/UniversalGraphicsTest_Foundation/Assets/ControlRenderScale.cs]: Also moved this file into the test directory, no need to be in the outermost directory

**Review by pbbastian — APPROVED** [review]

**Comment by esmelusina** [comment]
Would existing users expect any change in behavior?

**Comment by cdxntchou** [comment]
> Would existing users expect any change in behavior?

Only if they are running into the bug, in which case it will be fixed..  :)

**Review by esmelusina — APPROVED** [review]
LGTM.

**Pull request merged**